### PR TITLE
provider/aws: support custom endpoints for AWS EC2 ELB and IAM

### PR DESCRIPTION
**Pull request by turkenh**
Allows to specify AWS Endpoints for EC2, IAM and ELB directly from .tf file like: 

```
provider "aws" {
    access_key = "xxx"
    secret_key = "xxx"
    ec2_endpoint = "<ec2_endpoint_url>"
    elb_endpoint = "<elb_endpoint_url>"
    iam_endpoint = "<iam_endpoint_url>"
    insecure = "true"
    region = "us-east-1"
}
```
It also allows to disable HTTPS certificate check which is necessary for corporate solutions.

Solves https://github.com/hashicorp/terraform/issues/416 and https://github.com/hashicorp/terraform/issues/3974

### CODE REVIEW

**Comment by ecunal** [comment]
:+1:

**Comment by tugbabodrumlu** [comment]
:+1:

**Comment by kaskavalci** [comment]
:thumbsup:

**Comment by onuryilmaz** [comment]
:+1:

**Comment by SchwarzM** [comment]
wow this would be nice

**Comment by radeksimko** [comment]
Although I probably personally won't need this, I'm in favour of such feature, except I'd change the DSL.

Keeping all of these options in the 1st level doesn't look very scalable. Imagine how many endpoints this list may grow up to. I'd rather see something like this:

```hcl
provider "aws" {
    access_key = "xxx"
    secret_key = "xxx"
    endpoints {
        ec2 = "<ec2_endpoint_url>"
        elb = "<elb_endpoint_url>"
        iam = "<iam_endpoint_url>"
    }
    insecure = "true"
    region = "us-east-1"
}
```

File: builtin/providers/aws/config.go [hunk 1]
```diff
@@ -114,9 +120,21 @@ func (c *Config) Client() (interface{}, error) {
...
+				InsecureSkipVerify:true,
+			}
+		}
+
 		log.Println("[INFO] Initializing IAM Connection")
 		sess := session.New(awsConfig)
-		client.iamconn = iam.New(sess)
+
+		awsIamConfig := *awsConfig
+		awsIamConfig.Endpoint = aws.String(c.IamEndpoint)
```
Comment (radeksimko) [inline builtin/providers/aws/config.go]: Is it really optional? If it was, I guess a condition would be necessary here and below for each endpoint?

Comment (turkenh) [inline builtin/providers/aws/config.go]: If you do not set an endpoint, then, "c.IamEndpoint" is just an empty string here. And default value of config.Endpoint is also an empty string. So this is why I thought that there is no need for a condition here.

Comment (radeksimko) [inline builtin/providers/aws/config.go]: Ah, [you're right](https://github.com/aws/aws-sdk-go/blob/master/aws/config.go#L27-L28).
That's quite elegant solution then.

**Comment by turkenh** [comment]
@radeksimko Regarding your above comment about the change in DSL, I can change my pull request like that. But currently there are two endpoint options already implemented, namely dynamodb_endpoint and kinesis_endpoint at the first level. So, I should them try to move them under "endpoints" block then, right?

**Comment by radeksimko** [comment]
> currently there are two endpoint options already implemented, namely dynamodb_endpoint and kinesis_endpoint at the first level

Good point!

@catsby was merging #3255 and #2825 - what do you think about the DSL currently used VS proposed?

**Comment by afritzler** [comment]
+1

Any idea when this pull request will make it into the official release?

**Comment by catsby** [comment]
Hey Friends – 

Terribly sorry about the silence here. I agree with @radeksimko in that the top level will not be sustainable, and a block format like he suggested is preferable. 

We'd need to enable backwards compatibility with the current top level ones, however, so we'd have to support both.

If you could refactor to have at least the components you're suppling to use the block format, I can worry about the backwards compatibility. Or, if you're feeling _super_ generous, feel free to tackle it :smile: 

Thanks so much! I apologize again for the delay here. Let us know what you think

**Comment by turkenh** [comment]
@catsby and @radeksimko, I moved the endpoints added with this PR under "Endpoints" block. 
Did not touch previously added ones not to break something :)

**Comment by catsby** [comment]
Hey Friends – I just merged #5114 which is this PR but rebased on `master`, so I'm closing this. Thanks for all the work!

**Pull request closed without merging**